Initial C11 libatomic GCC builtin implementation

gyrovorbis opened the pull request:
This PR is to add the remaining GCC builtins required to enable all of C11's atomics, located in `<stdatomic.h>`. The majority of the builtins for primitive types are given to us by the compiler with `-matomic-model=soft-imask`; however, this evidently does not cover 64-bit (long long) primitive types nor does it cover the generic cases for atomic structures and buffers.

I've implemented these missing builtins, and I've implemented the 64-bit primitive back-ends using generic macros to generate the function calls, just in case we find anything else missing or wind up wanting to handle all primitive atomics by ourselves. 

- Initial implementation should cover everything C11 atomics entail, located at kernel/libc/c11/atomics.c

All of my preliminary tests are passing; however, I'm keeping this as a draft until I've created a KOS example/test scenario fully exercising all of C11 atomics.

gyrovorbis commented:
@ljsebald I was pondering some stuff last night and wanted to double check something with you real quick.

The C11 atomics API lets you query the implementation via `atomic_is_lock_free()` on an atomic to determine whether it "locks" or not. The default implementation for the stuff GCC is giving us via `-matomic-model=soft-imask` is all returning `true` there, so that model is clearly considered to be lock-free.

The question, though, is now that we're using spinlocks for the generic-case, is that still considered to be a "lock-free" implementation by the C standard, since it's technically non-blocking or do I need to modify that symbol to return `false` for the generic case?

ljsebald commented:
> @ljsebald I was pondering some stuff last night and wanted to double check something with you real quick.
> 
> The C11 atomics API lets you query the implementation via `atomic_is_lock_free()` on an atomic to determine whether it "locks" or not at runtime. The default implementation for the stuff GCC is giving us via `-matomic-model=soft-imask` is all returning `true` there, so that model is clearly considered to be lock-free.
> 
> The question, though, is now that we're using spinlocks for the generic-case, is that still considered to be a "lock-free" implementation by the C standard, since it's technically non-blocking or do I need to modify that symbol to return `false` for the generic case?

I'd imagine it wouldn't be proper to say that our implementation is lock free when it isn't. In the grand scheme of things, does it matter? Probably not... But to be proper, it should probably be the case that for any object that would trigger the spinlock case that it should say false for it being lock-free.

gyrovorbis commented:
> > @ljsebald I was pondering some stuff last night and wanted to double check something with you real quick.
> > The C11 atomics API lets you query the implementation via `atomic_is_lock_free()` on an atomic to determine whether it "locks" or not at runtime. The default implementation for the stuff GCC is giving us via `-matomic-model=soft-imask` is all returning `true` there, so that model is clearly considered to be lock-free.
> > The question, though, is now that we're using spinlocks for the generic-case, is that still considered to be a "lock-free" implementation by the C standard, since it's technically non-blocking or do I need to modify that symbol to return `false` for the generic case?
> 
> I'd imagine it wouldn't be proper to say that our implementation is lock free when it isn't. In the grand scheme of things, does it matter? Probably not... But to be proper, it should probably be the case that for any object that would trigger the spinlock case that it should say false for it being lock-free.

Sounds good. I am definitely wanting to do the right thing here.

1) updated back-end and example so that `atomic_is_lock_free()` returns false for our generic case
2) merged back with master
3) fixed a teensy little warning that I must've missed
4) reverified tests, nothing broke

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+   around memcpy() calls. */
+
+/* Size of each memory region covered by an individual lock. */
+#define GENERIC_LOCK_BLOCK_SIZE     (CPU_CACHE_BLOCK_SIZE * 4)
```
```suggestion
#define GENERIC_LOCK_BLOCK_SIZE     (CPU_CACHE_BLOCK_SIZE * 4)

```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+void __atomic_load(size_t size, 
+                   const volatile void *ptr, 
+                   void *ret, 
+                   int memorder)
+{
```
```suggestion
                   int memorder) {
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+void __atomic_store(size_t size, 
+                    volatile void *ptr, 
+                    void *val, 
+                    int memorder) 
+{
```
```suggestion
                    int memorder) {
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                       volatile void *ptr,
+                       void *val,
+                       void *ret,
+                       int memorder)
+{
```
```suggestion
                       int memorder) {
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                               void* expected,
+                               void* desired,
+                               int success_memorder,
+                               int fail_memorder)
+{
```
```suggestion
                               int fail_memorder) {
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                               int success_memorder,
+                               int fail_memorder)
+{
+    (void)success_memorder;
+    (void)fail_memorder;
+
+    bool retval;
+    const uintptr_t lock = address_to_spinlock(ptr);
```
```suggestion
    bool retval;
    const uintptr_t lock = address_to_spinlock(ptr);

    (void)success_memorder;
    (void)fail_memorder;
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+    if(memcmp((const void *)ptr, expected, size) == 0) {
+        memcpy((void *)ptr, desired, size);
+        retval = true;
+    } else {
```
```suggestion
    }
    else {
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                       void *ret,
+                       int memorder)
+{
+    (void)memorder;
+
+    const uintptr_t lock = address_to_spinlock(ptr);
```
```suggestion
    const uintptr_t lock = address_to_spinlock(ptr);

    (void)memorder;

```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                    void *val, 
+                    int memorder) 
+{
+    (void)memorder;
+    
+    const uintptr_t lock = address_to_spinlock(ptr);
```
```suggestion
    const uintptr_t lock = address_to_spinlock(ptr);

    (void)memorder;
    
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                   void *ret, 
+                   int memorder)
+{
+    (void)memorder;
+
+    const uintptr_t lock = address_to_spinlock(ptr);
```
```suggestion
    const uintptr_t lock = address_to_spinlock(ptr);

    (void)memorder;

```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+    __atomic_fetch_nand_##n(volatile void* ptr, \
+                            type val, \
+                            int memorder) { \
+        (void)memorder; \
+        const int irq = irq_disable(); \
+        type ret = *(type *)ptr; \
```
```suggestion
        const int irq = irq_disable(); \
        type ret = *(type *)ptr; \
        (void)memorder; \
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+    __atomic_fetch_##opname##_##n(volatile void* ptr, \
+                                  type val, \
+                                  int memorder) { \
+        (void)memorder; \
+        const int irq = irq_disable(); \
+        type ret = *(type *)ptr; \
```
```suggestion
        const int irq = irq_disable(); \
        type ret = *(type *)ptr; \
        (void)memorder; \
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+                                  type desired, \
+                                  bool weak, \
+                                  int success_memorder, \
+                                  int failure_memorder) \
+    { \
+        (void)weak; \
+        (void)success_memorder; \
+        (void)failure_memorder; \
+        const int irq = irq_disable(); \
+        bool retval; \
```
```suggestion
                                  int failure_memorder) { \
        const int irq = irq_disable(); \
        bool retval; \
        (void)weak; \
        (void)success_memorder; \
        (void)failure_memorder; \
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+#define ATOMIC_EXCHANGE_N_(type, n) \
+    type \
+    __atomic_exchange_##n(volatile void* ptr, type val, int model) { \
+        (void)model; \
+        const int irq = irq_disable(); \
+        const type ret = *(type *)ptr; \
```
```suggestion
        const int irq = irq_disable(); \
        const type ret = *(type *)ptr; \
        (void)model; \
```

ljsebald commented on kernel/libc/c11/atomics.c:
```diff
@@ -0,0 +1,221 @@
+#define ATOMIC_LOAD_N_(type, n) \
+    type \
+    __atomic_load_##n(const volatile void *ptr, int model) { \
+        (void)model; \
+        const int irq = irq_disable(); \
+        const type ret = *(type *)ptr; \
```
```suggestion
        const int irq = irq_disable(); \
        const type ret = *(type *)ptr; \
        (void)model; \
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+
+   /* Do several iterations worth of atomic operations for each thread. */
+   for(unsigned i = 0; i < ITERATION_COUNT; ++i) {
+
```
```suggestion
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+      bool expected = false;
+      while(!atomic_compare_exchange_weak(&bool_atomic,
+                                          &expected,
+                                          true))
+      { 
```
```suggestion
                                          true)) { 
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+         thrd_yield();
+         expected = false;     
+      }
+      printf("Thread[%2u]: Acquired atomic bool lock.\n", tid);
```
```suggestion

      printf("Thread[%2u]: Acquired atomic bool lock.\n", tid);
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+      expected = true;
+      if(!atomic_compare_exchange_strong(&bool_atomic,
+                                         &expected,
+                                         false))
+      {
```
```suggestion
                                         false)) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   /* Create N-1 threads running the "thread()" function to
+      manipulate and fight over various atomic data. */ 
+   for(unsigned t = 0; t < THREAD_COUNT - 1; ++t) {
+      if(thrd_create(&threads[t], thread, (void *)t + 1)
+            != thrd_success) 
+      {
```
```suggestion
      if(thrd_create(&threads[t], thread,
                     (void *)t + 1) != thrd_success) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   if(atomic_flag_test_and_set(&flag_atomic)) {
+      fprintf(stderr, "flag_atomic left in unexpected state: [true]\n");
+      retval = -1;
+   } else printf("\tFlag atomics work.\n");
```
```suggestion
   }
   else {
       printf("\tFlag atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   bool expected = false;
+   if(!atomic_compare_exchange_weak(&bool_atomic,
+                                    &expected,
+                                    true))
+   {
```
```suggestion
                                    true)) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   for(unsigned i = 0; i < THREAD_COUNT; ++i)
+      byte_expected_value |= i;
+
+   if((byte_value = atomic_load(&byte_atomic))
+         != byte_expected_value)
+   {
```
```suggestion
   if((byte_value = atomic_load(&byte_atomic)) != byte_expected_value) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+              "byte_atomic left in unexpected state: [%u]\n", 
+              byte_value);
+      retval = -1;
+   } else printf("\t8-bit atomics work.\n");
```
```suggestion
   }
   else {
       printf("\t8-bit atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+              "short_atomic left in unexpected state: [%i]\n", 
+              short_value);
+      retval = -1;
+   } else printf("\t16-bit atomics work.\n");
```
```suggestion
   }
   else {
       printf("\t16-bit atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+              "int_atomic left in unexpected state: [%d]\n", 
+              int_value);
+      retval = -1;
+   } else printf("\t32-bit atomics work.\n");
```
```suggestion
   }
   else {
       printf("\t32-bit atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+              "longlong_atomic left in unexpected state: [%llu]\n", 
+              longlong_value);
+      retval = -1;
+   } printf("\t64-bit atomics work.\n");
```
```suggestion
   }
   else {
       printf("\t64-bit atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+              "ptrdiff_atomic left in unexpected state: [%d]\n", 
+              ptrdiff_value);
+      retval = -1;
+   } else printf("\tptrdiff_t atomics work.\n");
```
```suggestion
   }
   else {
       printf("\tptrdiff_t atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   {
+      fprintf(stderr, "bool_atomic left in unexpected state: [true]\n");
+      retval = -1;
+   } else printf("\tBool atomics work.\n");
```
```suggestion
   }
   else {
       printf("\tBool atomics work.\n");
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   for(int i = 0; i < THREAD_COUNT; ++i)
+      int_expected_value &= i;
+   
+   if((int_value = atomic_load(&int_atomic)) 
+         != int_expected_value)
+   {
```
```suggestion
   if((int_value = atomic_load(&int_atomic)) != int_expected_value) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+
+   /* Verify atomic long long state. */
+   uint64_t longlong_value;
+   if((longlong_value = atomic_load(&longlong_atomic)) 
+         != THREAD_COUNT * ITERATION_COUNT) 
+   {
```
```suggestion
   if((longlong_value = atomic_load(&longlong_atomic)) !=
           THREAD_COUNT * ITERATION_COUNT) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+   /* Verify atomic short state. */
+   short short_value;
+   /* Note that memory order shouldn't do anything for our platform, just testing. */
+   if((short_value = atomic_load(&short_atomic))
+         != -(THREAD_COUNT * ITERATION_COUNT))
+   {
```
```suggestion
   if((short_value = atomic_load(&short_atomic)) !=
           -(THREAD_COUNT * ITERATION_COUNT)) {
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+
+   /* Verify atomic int state. */
+   int int_value, int_expected_value = INT_MAX;
+   for(int i = 0; i < THREAD_COUNT; ++i)
+      int_expected_value &= i;
```
```suggestion
   for(int i = 0; i < THREAD_COUNT; ++i) {
       int_expected_value &= i;
   }
```

ljsebald commented on examples/dreamcast/basic/threading/atomics/atomics.c:
```diff
@@ -0,0 +1,388 @@
+
+   /* Verify atomic byte state. */
+   uint8_t byte_value, byte_expected_value = 0;
+   for(unsigned i = 0; i < THREAD_COUNT; ++i)
+      byte_expected_value |= i;
```
```suggestion
   for(unsigned i = 0; i < THREAD_COUNT; ++i) {
       byte_expected_value |= i;
   }
```

ljsebald commented on examples/dreamcast/basic/threading/Makefile:
```diff
@@ -1,7 +1,7 @@
 #
 # examples/dreamcast/basic/threading/Makefile
-# (c)2002 Megan Potter
+# (c) 2002 Megan Potter
```
```suggestion
# Copyright (C) 2002 Megan Potter
```

gyrovorbis commented:
Alrighty,  addressed review changes and retested, because I'm always paranoid. Looking good!

![Screenshot from 2023-09-15 02-14-39](https://github.com/KallistiOS/KallistiOS/assets/5545520/c46b3ee2-ad98-4383-914d-5ecae2cfcc0e)
